openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +39,71 @@ def process_request(self, request):
+            # .. setting_warning:  Must be smaller than SESSION_INACTIVITY_TIMEOUT_IN_SECONDS.
+            frequency_time_in_seconds = getattr(settings, "SESSION_ACTIVITY_SAVE_DELAY_SECONDS", 900)
             # what time is it now?
             utc_now = datetime.utcnow()
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    # Convert the ISO string back to a datetime object
+                    last_touch = datetime.fromisoformat(last_touch_str)
+
+                    # compute the delta since last time user came to the server
+                    time_since_last_activity = utc_now - last_touch
+
+                    # did we exceed the timeout limit?
+                    if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
+                        # yes? Then log the user out
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    LOG.debug("Parsing last touch time failed: %s", e)
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                # Enable NR tracing for this view based on course
+                monitoring_utils.set_custom_attribute('session_inactivity.first_login', True)
+                LOG.debug("No previous activity timestamp found (first login)")
+
+            # Store activity timestamp
+            request.session[LAST_TOUCH_KEYNAME] = utc_now.isoformat()
+
+            # Periodically allow a full save (every n seconds)
+            last_save = request.session.get(LAST_SESSION_SAVE_TIME_KEYNAME)
+            current_time = datetime.utcnow().isoformat()
+
+            monitoring_utils.set_custom_attribute('session_inactivity.activity_seen', current_time)
+            if not last_save or (
+                datetime.fromisoformat(last_save) + timedelta(seconds=frequency_time_in_seconds) < datetime.utcnow()
+            ):
```

Comment:
[Note to self:]
- This is where I need to pick up the review.
- The code/comments weren't immediately clear, but no changes are being requested until I can continue.